tests/unit/fem/test_pa_grad.cpp

```diff
@@ -0,0 +1,254 @@
+   int size;
+   MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+   Mesh * smesh;
+   if (dim == 2)
+   {
+      smesh = new Mesh(num_elements, num_elements, Element::QUADRILATERAL, true);
+   }
+   else
+   {
+      smesh = new Mesh(num_elements, num_elements, num_elements,
+                       Element::HEXAHEDRON, true);
+   }
+   ParMesh * mesh = new ParMesh(MPI_COMM_WORLD, *smesh);
+   delete smesh;
+   FiniteElementCollection *h1_fec = new H1_FECollection(order, dim);
+   FiniteElementCollection *nd_fec = new ND_FECollection(order, dim);
+   ParFiniteElementSpace h1_fespace(mesh, h1_fec);
+   ParFiniteElementSpace nd_fespace(mesh, nd_fec);
+
+   ParDiscreteLinearOperator assembled_grad(&h1_fespace, &nd_fespace);
+   assembled_grad.AddDomainInterpolator(new GradientInterpolator);
+   const int skip_zeros = 1;
+   assembled_grad.Assemble(skip_zeros);
+   assembled_grad.Finalize(skip_zeros);
+   HypreParMatrix * assembled_grad_mat = assembled_grad.ParallelAssemble();
+
+   ParDiscreteLinearOperator pa_grad(&h1_fespace, &nd_fespace);
+   pa_grad.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   pa_grad.AddDomainInterpolator(new GradientInterpolator);
+   pa_grad.Assemble();
+   OperatorPtr pa_grad_oper;
+   pa_grad.FormRectangularSystemMatrix(pa_grad_oper);
+
+   int insize, outsize;
+   if (transpose)
+   {
+      insize = assembled_grad_mat->Height();
+      outsize = assembled_grad_mat->Width();
+   }
+   else
+   {
+      insize = assembled_grad_mat->Width();
+      outsize = assembled_grad_mat->Height();
+   }
+   Vector xv(insize);
+   Vector assembled_y(outsize);
+   Vector pa_y(outsize);
+   assembled_y = 0.0;
+   pa_y = 0.0;
+
+   xv.Randomize();
+   if (transpose)
+   {
+      assembled_grad_mat->MultTranspose(xv, assembled_y);
+      pa_grad_oper->MultTranspose(xv, pa_y);
+   }
+   else
+   {
+      assembled_grad_mat->Mult(xv, assembled_y);
+      pa_grad_oper->Mult(xv, pa_y);
+   }
+
+   Vector error_vec(pa_y);
+   error_vec -= assembled_y;
+   // serial norms and serial error; we are enforcing equality on each processor
+   // in the test
+   double error = error_vec.Norml2() / assembled_y.Norml2();
+
+   for (int p = 0; p < size; ++p)
+   {
+      if (rank == p)
+      {
+         std::cout << "[" << rank << "]";
+         // std::cout << "pa_y.Norml2() = " << pa_y.Norml2() << std::endl;
+         // std::cout << "assembled_y.Norml2() = " << assembled_y.Norml2() << std::endl;
+         std::cout << "[par] dim " << dim << " ne " << num_elements << " order "
+                   << order;
+         if (transpose)
+         {
+            std::cout << " T";
+         }
+         std::cout << ": error in PA gradient: " << error << std::endl;
+         std::cout.flush();
+      }
+      MPI_Barrier(MPI_COMM_WORLD);
+   }
+
+   delete h1_fec;
+   delete nd_fec;
+   delete assembled_grad_mat;
+   delete mesh;
+
+   return error;
+}
+
+TEST_CASE("ParallelPAGradient", "[Parallel], [ParallelPAGradient]")
+{
+   for (bool transpose : {false, true})
```

Comment:
I suggest using `GENERATE` here instead of for loops (see my comment in `test_pa_idinterp.cpp`)